File: common/src/main/java/dev/throwouterror/game/common/Transform.java
/*
 * Copyright (c) Creepinson
 */

package dev.throwouterror.game.common;


import org.joml.Matrix4f;
import org.joml.Vector3f;

import java.io.Serializable;

public class Transform implements Serializable {
    private final Vector3f position;
    private final Matrix4f rotation;
    private final Vector3f scale;

    public Transform() {
        this(new Vector3f(0, 0, 0), new Matrix4f(), new Vector3f(1, 1, 1));
    }

    public Transform(Vector3f position, Matrix4f rotation, Vector3f scale) {
        this.position = position;
        this.rotation = rotation;
        this.scale = scale;
    }

    public static Transform pos(float x, float y, float z) {
        return new Transform(new Vector3f(x, y, z), new Matrix4f(), new Vector3f(1, 1, 1));
    }

    public Transform clone() {
        return new Transform(new Vector3f(this.position), new Matrix4f(this.rotation), new Vector3f(this.scale));
    }

    public Vector3f getPosition() {
        return this.position;
    }

    public Matrix4f getRotation() {
        return this.rotation;
    }

    public Vector3f getScale() {
        return this.scale;
    }

    @Override
    public String toString() {
        return "Transform{" +
                "position=" + position +
                ", rotation=" + rotation +
                ", scale=" + scale +
                '}';
    }
}

File: common/src/main/kotlin/dev/throwouterror/game/common/Transform.kt
package dev.throwouterror.game.common

import dev.throwouterror.util.data.JsonObjectBuilder
import dev.throwouterror.util.data.JsonUtils
import dev.throwouterror.util.math.Direction
import dev.throwouterror.util.math.Tensor
import java.io.Serializable

class Transform @JvmOverloads constructor(
    val position: Tensor = Tensor.VECTOR_ZERO.clone(),
    val rotation: Tensor = Tensor.VECTOR_ZERO.clone(),
    val scale: Tensor = Tensor(1.0, 1.0, 1.0)
) : Serializable {

    fun clone(): Transform {
        return Transform(this.position.clone(), this.rotation.clone(), this.scale.clone())
    }

    override fun toString(): String {
        return "Transform{" +
            "position=" + position +
            ", rotation=" + rotation +
            ", scale=" + scale +
            '}'.toString()
    }

    fun toJson(): String {
        return JsonObjectBuilder().json {
            "position" to position
        }.toString()
    }

    fun move(t: Tensor): Transform {
        this.position.add(t)
        return this
    }

    fun move(t: Direction): Transform {
        return this.move(t.facingVec)
    }

    companion object {
        fun fromJson(json: String): Transform {
            return JsonUtils.builder.fromJson(json, Transform::class.java)
        }

        fun pos(x: Double, y: Double, z: Double): Transform {
            return Transform(Tensor(x, y, z), Tensor.VECTOR_ZERO.clone(), Tensor(1.0, 1.0, 1.0))
        }

        fun pos(x: Float, y: Float, z: Float): Transform {
            return pos(x.toDouble(), y.toDouble(), z.toDouble())
        }

        fun pos(x: Int, y: Int, z: Int): Transform {
            return pos(x.toDouble(), y.toDouble(), z.toDouble())
        }
    }
}
